### PR TITLE
Issue #117 - Add Parenthesis Hugging

### PR DESCRIPTION
**Pull request by mstackhouse**
This PR incorporates all necessary updates to satisfy #117. Doc updates and unit testing updates have been satisfied. 

Unit tests in test-num_fmt.R give a fuller example of the functionality, but a few examples of the functionality are as follows:

```r
load('vignettes/adlb.Rdata')

tplyr_table(adlb, TRTA, PARAMCD == "CK") %>% 
  add_layer(
    group_desc(AVAL, by=vars(PARAMCD, AVISIT)) %>% 
      set_format_strings(
        TEST = f_str("xxx.x (XXX.x)", mean, sd, empty="NA")
      ) %>% 
      set_precision_by(PARAMCD)
  ) %>% 
  build() %>% 
  head() %>% 
  select(-starts_with('ord'))

#> # A tibble: 3 × 6
#>   row_label1 row_label2 row_label3 var1_Placebo  `var1_Xanomeline High Dose`
#>   <chr>      <chr>      <chr>      <chr>         <chr>                      
#> 1 CK         Week 12    TEST       140.0 (148.6) "140.1 (115.5)"            
#> 2 CK         Week 24    TEST       246.6 (438.5) " 55.5   (3.5)"            
#> 3 CK         Week 8     TEST       116.0  (78.9) " 93.3  (80.6)"            
#> # … with 1 more variable: `var1_Xanomeline Low Dose` <chr>
```

```r
tplyr_table(adlb, TRTA, PARAMCD == "CK") %>% 
  add_layer(
    group_desc(AVAL, by=vars(PARAMCD, AVISIT)) %>% 
      set_format_strings(
        TEST = f_str("a.a (A.a)", mean, sd, empty="NA")
      ) %>% 
      set_precision_by(PARAMCD)
  ) %>% 
  build() %>% 
  head() %>% 
  select(-starts_with('ord'))

#> # A tibble: 3 × 6
#>   row_label1 row_label2 row_label3 var1_Placebo `var1_Xanomeline High Dose`
#>   <chr>      <chr>      <chr>      <chr>        <chr>                      
#> 1 CK         Week 12    TEST       " 140 (149)" " 140 (115)"               
#> 2 CK         Week 24    TEST       " 247 (438)" "  56   (4)"               
#> 3 CK         Week 8     TEST       " 116  (79)" "  93  (81)"               
#> # … with 1 more variable: `var1_Xanomeline Low Dose` <chr>
```

```r
tplyr_table(adae, TRTA) %>%
  add_layer(
    group_count(AEDECOD) %>%
      set_distinct_by(USUBJID) %>%
      set_format_strings(f_str("xxx (XXX.xx%) [A]", distinct_n, distinct_pct, n))
  ) %>%
  build() %>%
  head() %>% 
  select(row_label1, `var1_Xanomeline Low Dose`)

#> # A tibble: 6 × 2
#>   row_label1         `var1_Xanomeline Low Dose`
#>   <chr>              <chr>                     
#> 1 ACTINIC KERATOSIS  "  0   (0.00%)  [0]"      
#> 2 ALOPECIA           "  0   (0.00%)  [0]"      
#> 3 BLISTER            "  5  (11.90%)  [8]"      
#> 4 COLD SWEAT         "  0   (0.00%)  [0]"      
#> 5 DERMATITIS ATOPIC  "  0   (0.00%)  [0]"      
#> 6 DERMATITIS CONTACT "  1   (2.38%)  [2]"
```

Some other notes:

- This required restructuring of the `f_str()$settings` parameter, so all of those unit tests had to be updated but I verified that the results all translated
- I moved some code around to different files. format.R specifically was too large so I split it into format.R, num_fmt.R, and set_format_strings.R
  - I also updated unit testing to reflect this, though set_format_strings is tested through essentially every other build file, and num_fmt is dedicated to the hugging specific updates because general format testing exists elsewhere
- I split the descriptive statistics layer vignette to pull empty values, auto precision, external precision, and the new docs around parenthesis hugging into a new advanced desc layer vignette because it was far too long (vignettes generally need some reorganization as well).

### CODE REVIEW

**Comment by mstackhouse** [comment]
@elimillera  when we can approve this update, I want to do a CRAN push as version 1.1.0, which is why I updated the version number.

**Review by sadchla-codes — COMMENTED** [review]
Besides the comments/questions that I made, I think this looks great!

**Comment by mstackhouse** [comment]
@sadchla-codes  I just added a new function - could you take a look? It's `apply_conditional_formats()` and hopefully the function doc explains it well.

In practice, I see it being used as follows:

```r
load(test_path('adsl.Rdata'))

t <- tplyr_table(adsl, TRT01P) %>% 
  add_layer(
    group_count(RACE, by="Race n (%)") %>% 
      set_format_strings(f_str("xx (XX.x%)", n, pct))
  )

dat <- t %>% 
  build() 

# Replace whole string
dat %>% 
  mutate(
    across(starts_with('var1'), ~ apply_conditional_format(., 2, x == 0, " 0        ", full_string=TRUE))
  )

# Replace partial string
dat %>% 
  mutate(
    across(starts_with('var1'), ~ apply_conditional_format(., 2, x < 1, "(<1%)"))
  )

# Other use of across to demonstrate that non-numeric character strings are simply ignored
dat %>% 
  mutate(
    across(where(is.character), ~ apply_conditional_format(., 2, x < 1, "(<1%)"))
  )
```

File: man/str_extractors.Rd [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+\description{
+These functions allow you to extract segments of information from within a
+result string by targetting specific format groups. \code{str_extract_fmt_group()}
+allows you to pull out the individual format group string, while
+\code{str_extract_num()} allows you to pull out that specific numeric result.
+}
+\details{
+Format groups refer to individual segments of a string. For example, given
+the string ' 5 (34.4\%) \link{9}', there are three separate format groups, which
+are ' 5', '(34.4\%)', and '\link{9}'.
```
Comment (sadchla-codes) [inline man/str_extractors.Rd:30]: This is a warning from the R CMD check related to `\link{9}` (picture below)
tplyr-failed

Comment (mstackhouse) [reply man/str_extractors.Rd:30]: @sadchla-codes  fixed this one

**Review by sadchla-codes — APPROVED** [review]

**Review by elimillera — COMMENTED** [review]
Finished with the actual code review, working on functionality and things next

**Comment by mstackhouse** [comment]
@sadchla-codes  sorry... I added new vignettes... Can you do some proofreading for me?

**Comment by sadchla-codes** [comment]
@mstackhouse I'm on it!

**Review by sadchla-codes — APPROVED** [review]

**Review by elimillera — APPROVED** [review]
Looks really good. I haven't found a way to break it.

**Pull request merged**